bee-signing/src/ternary/mss.rs

```diff
@@ -0,0 +1,407 @@
+    type Error = Error;
+
+    fn verify(&self, message: &Trits<T1B1>, signature: &Self::Signature) -> Result<bool, Self::Error> {
+        if signature.size() % SIGNATURE_FRAGMENT_LENGTH != 0 || signature.size() < 2 * SIGNATURE_FRAGMENT_LENGTH {
+            return Err(Error::InvalidSignatureSize);
+        }
+
+        let depth = self.depth.ok_or(Error::MissingDepth)?;
+        let signature_index = signature.index.ok_or(Error::MissingSignatureIndex)?;
+
+        let mut sponge = S::default();
+        let underlying_signature = K::Signature::from_trits(
+            signature.state[0..((signature.state.len() / SIGNATURE_FRAGMENT_LENGTH) - 1) * SIGNATURE_FRAGMENT_LENGTH]
+                .to_buf(),
+        )
+        .map_err(|_| Error::FailedUnderlyingSignatureGeneration)?;
+
+        // Safe to unwrap since we already validated the signature size.
+        let siblings: &Trits<T1B1> = signature.state.chunks(SIGNATURE_FRAGMENT_LENGTH).last().unwrap();
+        // Recovers underlying public key from underlying signature.
+        let underlying_public_key = underlying_signature
+            .recover_public_key(message)
+            .map_err(|_| Self::Error::FailedUnderlyingPublicKeyRecovery)?;
+        let mut hash = TritBuf::<T1B1Buf>::zeros(HASH_LENGTH);
+
+        hash.copy_from(underlying_public_key.as_trits());
+
+        // Hash the underlying public key with the merkle branch to recover the merkle root.
+        let mut j = 1;
+        for (i, sibling) in siblings.chunks(HASH_LENGTH).enumerate() {
+            if depth - 1 == i as u8 {
+                break;
+            }
+
+            if signature_index & j != 0 {
+                sponge.absorb(sibling).map_err(|_| Self::Error::FailedSpongeOperation)?;
+                sponge.absorb(&hash).map_err(|_| Self::Error::FailedSpongeOperation)?;
+            } else {
+                sponge.absorb(&hash).map_err(|_| Self::Error::FailedSpongeOperation)?;
+                sponge
+                    .absorb(&sibling)
+                    .map_err(|_| Self::Error::FailedSpongeOperation)?;
+            }
+            sponge
+                .squeeze_into(&mut hash)
+                .map_err(|_| Self::Error::FailedSpongeOperation)?;
+            sponge.reset();
+
+            j <<= 1;
+        }
+
+        Ok(hash == self.state)
+    }
+
+    fn size(&self) -> usize {
+        self.state.len()
+    }
+
+    fn from_trits(state: TritBuf<T1B1Buf>) -> Result<Self, Self::Error> {
+        Ok(Self {
+            state,
+            depth: None,
+            sponge: PhantomData,
+            key: PhantomData,
+        })
+    }
+
+    fn as_trits(&self) -> &Trits<T1B1> {
+        &self.state
+    }
+}
+
+/// Merkle Signature Scheme signature.
+pub struct MssSignature<S> {
+    state: TritBuf<T1B1Buf>,
+    index: Option<u64>,
+    sponge: PhantomData<S>,
+}
+
+impl<S: Sponge + Default> MssSignature<S> {
+    /// Set the index of the signature.
+    pub fn index(mut self, index: u64) -> Self {
```

Comment:
I think this method name might be a little misleading. Single-noun methods are usually getters in Rust

Reply:
So, `set_index` ?

Reply:
Or `with_index` ?

Reply:
Done